sdk/nodejs/dynamic/index.ts

```diff
@@ -184,7 +186,15 @@ export abstract class Resource extends resource.CustomResource {
         if (props[providerKey]) {
             throw new Error("A dynamic resource must not define the __provider key");
         }
-        props[providerKey] = serializeProvider(provider);
+        let serializedProvider: Promise<string>;
+        if (providerCache.has(provider)) {  
```

Comment:
Will this be valid if the implementations of the various functions that implement the resource provider have changed? I'm thinking in particular of the functions being reassigned to new implementations.

Reply:
Unfortunately, there are endless ways that this could be invalid. For instance, even if you don't mutate the functions you could increment some shared global variable. Through that lens, this is technically a breaking change although one that seems less likely for users to hit. Through that lense I see a few options:

1. accept this change, but extend the diffing logic to check equality of each of the provider functions. This still has issues as noted, but probably covers the 99.99%
2. Make this behavior opt-in (maybe via `dynamic.CacheableResourceProvider`). I don't like this option as this behavior should really be the default. 

Thoughts? I suppose we could do (2), and cut over to (1) as a part of 3.0.

Reply:
> Unfortunately, there are endless ways that this could be invalid. For instance, even if you don't mutate the functions you could increment some shared global variable. Through that lens, this is technically a breaking change although one that seems less likely for users to hit.

This is a good point. FWIW, we probably haven't documented that closing over mutable global state in a dynamic provider is generally a bad idea, since the provider won't observe any mutations once it is serialized.

> Thoughts? I suppose we could do (2), and cut over to (1) as a part of 3.0.

This seems good to me, though I would use an envvar for the opt-in (despite the big hammer, it doesn't require code changes).

Honestly, I'm probably being a little too careful with this particular aspect of the change. @lukehoban wdyt?

Reply:
FWIW - I'm actually not sure this cache will help in many/most cases.  It's pretty common style to do something like `super(new MyResourceProvider(), ...)` or equivalent.  In fact, of the 5 examples of TypeScript dynamic providers in pulumi/examples, I'm pretty sure only one of them would see any benefit from this cache.  

As for the breaking change aspect of this - I think given we are so close to 3.0, and that this isn't a particularly practically severe issue (4k dynamic resources is meaningfully outside of current "normal"), we could just choose to only deliver this side of the change into 3.0, and take it as a documented breaking change.

> FWIW, we probably haven't documented that closing over mutable global state in a dynamic provider is generally a bad idea, since the provider won't observe any mutations once it is serialized.

We should probably add a paragraph to the dynamic providers docs on this point, both in terms of the impact that closure serialization has here, and that this caching has.

Reply:
Another thing to note here is that as long as serialization is deterministic, we'll get cache hits on the provider side even if they use the pattern you describe @lukehoban.